pkg/core/native/crypto.go

```diff
@@ -244,6 +252,72 @@ func (c *Crypto) verifyWithEd25519(_ *interop.Context, args []stackitem.Item) st
 	return stackitem.NewBool(ed25519.Verify(pubkey, msg, signature))
 }
 
+func (c *Crypto) recoverSecp256K1(_ *interop.Context, args []stackitem.Item) stackitem.Item {
+	msgH, err := args[0].TryBytes()
+	if err != nil {
+		panic(fmt.Errorf("invalid message hash stackitem: %w", err))
+	}
+	if len(msgH) != 32 {
+		return stackitem.Null{}
+	}
+	signature, err := args[1].TryBytes()
+	if err != nil {
+		panic(fmt.Errorf("invalid signature stackitem: %w", err))
+	}
+	if len(signature) != 64 && len(signature) != 65 {
+		return stackitem.Null{}
+	}
+
+	// ecdsa library (as most other general purpose crypto libraries) work only with
+	// canonical representation, compact representation is a pure Ethereum feature (ref.
+	// https://eips.ethereum.org/EIPS/eip-2098#specification). Also, ecdsa expects
+	// leading signature byte to be a key recovery ID.
+	canonical := koblitzSigToCanonical(signature)
+	pub, _, err := ecdsa.RecoverCompact(canonical, msgH)
+	if err != nil {
+		return stackitem.Null{}
+	}
+	return stackitem.NewByteArray(pub.SerializeCompressed())
+}
+
+// koblitzSigToCanonical converts compact Secp256K1 signature representation
+// (https://eips.ethereum.org/EIPS/eip-2098#specification) to canonical
+// form (https://www.secg.org/sec1-v2.pdf, section 4.1.6) and moves key recovery
+// ID from the last byte to the first byte of the resulting signature (for both
+// compact and non-compact forms), as it is required by ecdsa package.
+func koblitzSigToCanonical(signature []byte) []byte {
+	// Convert from compact input format (r[32] || yParityAndS[32] where yParity
+	// is fused into the top bit of s) to: r[32] || s[32] || v[1].
+	if len(signature) == 64 {
+		s := slices.Clone(signature[32:]) // don't modify the original slice.
+		yParity := s[0] >> 7
+		s[0] = s[0] & ((1 << 7) - 1)
+		res := make([]byte, 65)
+		copy(res[:32], signature[:32])
+		copy(res[32:64], s)
+		if yParity == 0 {
+			res[64] = 27 // compact key recovery code for uncompressed public key inherited from Bitcoin.
+		} else {
+			res[64] = 28 // compact key recovery code for compressed public key inherited from Bitcoin.
+		}
+		signature = res
```

Comment:
@roman-khimov, I can move these magic numbers to package-level constants if it will help. `decred/dcrd` declares some of them, but they are unexported.

Reply:
I'm OK with these numbers, they're irrelevant in other places anyway. But you're allocating twice for 64-byte case (and doing an additional copy), this better be optimized out.

Reply:
Sure, fixed.